build_pros.cmake

```diff
@@ -0,0 +1,235 @@
+# ======== Policies ========
+
+if(POLICY CMP0075) # Include file check macros honor CMAKE_REQUIRED_LIBRARIES, CMake >= 3.12
+    cmake_policy(SET CMP0075 NEW)
+endif()
+
+if(POLICY CMP0077) # Cache variables override since 3.12
+    cmake_policy(SET CMP0077 NEW)
+endif()
+
+# ======== Includes ========
+
+include(CheckSymbolExists)
+include(CheckFunctionExists)
+include(CheckIncludeFile)
+include(CMakePushCheckState)
+include(CheckCSourceCompiles)
+include(CheckCCompilerFlag)
+include(TestBigEndian)
+
+# If platform is Emscripten
+if(${CMAKE_SYSTEM_NAME} STREQUAL "Emscripten")
+    set(EMSCRIPTEN 1)
+endif()
+
+
+# ======== Optimizations and warnings ========
+
+# Strip garbage
+if(APPLE)
+    string(REGEX REPLACE "-O3" ""
+        CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE}")
+    set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -O2")
+    set(LINK_FLAGS_RELEASE  "${LINK_FLAGS_RELEASE} -dead_strip")
+elseif(NOT MSVC)
+    if(EMSCRIPTEN)
+        set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -Os -fdata-sections -ffunction-sections")
+        if("${CMAKE_C_COMPILER_ID}" STREQUAL "Clang")
+            set(LINK_FLAGS_RELEASE  "${LINK_FLAGS_RELEASE} -dead_strip")
+        else()
+            set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -Wl,--gc-sections -Wl,-s")
+            set(LINK_FLAGS_RELEASE  "${LINK_FLAGS_RELEASE} -Wl,--gc-sections -Wl,-s")
+        endif()
+    else()
+        string(REGEX REPLACE "-O3" ""
+            CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE}")
+        set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -O2 -fdata-sections -ffunction-sections")
```

Comment:
This whole block in the `else()` section: Are any of them _truly_ needed?
The autotools side don't do any of these.

Reply:
I use these flags at my side to remove the dead code while building the stuff. Okay, Do you suggest to apply these flags at externally-passed `CMAKE_C_FLAGS` instead?

Reply:
My opinion: These needn't be necessary, so I vote that don't add any unnecessary
complexities, not at the moment.  _However_: if we do like these options, then they
must be added to autotools side, too.

Reply:
I made the test, so, look:

Binaries with optimizations block:
![Снимок экрана от 2021-07-19 17-49-58](https://user-images.githubusercontent.com/6751442/126180007-81a2371d-1fc6-4f7a-b7f2-089283634056.png)

And the same binaries once I removed the whole optimizations block:
![Снимок экрана от 2021-07-19 17-50-43](https://user-images.githubusercontent.com/6751442/126180065-1d855a4a-3208-4f6b-ac32-ea6291cb34ad.png)

You can see that optimized binaries have the slightly smaller size

HOWEVER, that will happen when I will use the `CMAKE_BUILD_TYPE=MinSizeRel` and without optimizations block:
![Снимок экрана от 2021-07-19 17-53-01](https://user-images.githubusercontent.com/6751442/126180291-80fbd43f-205b-4e9b-b4d0-f8884de48ff0.png)

Reply:
Anyway, while the MinSizeRel exists, this block has no sense.

Reply:
With older gcc, I get the following at every invocation of gcc:
`i686-pc-mingw32-gcc: --gc-sections: linker input file unused because linking not done`
Or: `gcc34: --gc-sections: linker input file unused because linking not done`

Reply:
So, the next time, at my projects, I'll add these checks to prevent similar errors while using these flags. At here I completely removed this block as it has no sense because of the MinSizeRel.